src/org/opensolaris/opengrok/history/MercurialRepository.java

```diff
@@ -161,28 +162,35 @@ Executor getHistoryLogExecutor(File file, String sinceRevision)
         cmd.add(RepoCommand);
         cmd.add("log");
 
-        // For plain files we would like to follow the complete history
-        // (this is necessary for getting the original name in given revision
-        // when handling renamed files)
-        if (!file.isDirectory()) {
-            cmd.add("-f");
-        }
-
-        // If this is non-default branch we would like to get the changesets
-        // on that branch and also any changesets from the parent branch(es).
-        if (sinceRevision != null) {
-            cmd.add("-r");
-            String[] parts = sinceRevision.split(":");
-            if (parts.length == 2) {
-                cmd.add("reverse(" + parts[0] + "::'" + getBranch() + "')");
+        if (file.isDirectory()) {
+            // If this is non-default branch we would like to get the changesets
+            // on that branch and also follow any changesets from the parent branch.
+            if (sinceRevision != null) {
+                cmd.add("-r");
+                String[] parts = sinceRevision.split(":");
+                if (parts.length == 2) {
+                    cmd.add("reverse(" + parts[0] + "::'" + getBranch() + "')");
+                } else {
+                    throw new HistoryException(
+                            "Don't know how to parse changeset identifier: "
+                            + sinceRevision);
+                }
             } else {
-                throw new HistoryException(
-                        "Don't know how to parse changeset identifier: "
-                        + sinceRevision);
+                cmd.add("-r");
+                cmd.add("reverse(0::'" + getBranch() + "')");
             }
         } else {
-            cmd.add("-r");
-            cmd.add("reverse(0::'" + getBranch() + "')");
+            // For plain files we would like to follow the complete history
+            // (this is necessary for getting the original name in given revision
+            // when handling renamed files)
+            // It is not needed to filter on a branch as 'hg log' will follow
+            // the active branch.
+            // Due to behavior of recent Mercurial versions, it is not possible
+            // to filter the changesets of a file based on revision.
+            // For files this does not matter since if getHistory() is called
+            // for a file, the file has to be renamed so we want its complete history.
```

Comment:
> For files this does not matter since if getHistory() is called
for a file, the file has to be renamed so we want its complete history.

This should be part of the javadoc for this function